pkg/operator/cassandra/controller/controller.go

```diff
@@ -0,0 +1,304 @@
+		podLister:                  podInformer.Lister(),
+		podListerSynced:            podInformer.Informer().HasSynced,
+		serviceLister:              serviceInformer.Lister(),
+		serviceListerSynced:        serviceInformer.Informer().HasSynced,
+		serviceAccountLister:       serviceAccountInformer.Lister(),
+		serviceAccountListerSynced: serviceAccountInformer.Informer().HasSynced,
+		roleLister:                 roleInformer.Lister(),
+		roleListerSynced:           roleInformer.Informer().HasSynced,
+		roleBindingLister:          roleBindingInformer.Lister(),
+		roleBindingListerSynced:    roleBindingInformer.Informer().HasSynced,
+
+		queue:    workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), clusterQueueName),
+		recorder: recorder,
+	}
+
+	// Add event handling functions
+
+	clusterInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			newCluster := obj.(*cassandrav1alpha1.Cluster)
+			cc.enqueueCluster(newCluster)
+		},
+		UpdateFunc: func(old, new interface{}) {
+			newCluster := new.(*cassandrav1alpha1.Cluster)
+			oldCluster := old.(*cassandrav1alpha1.Cluster)
+			if newCluster.ResourceVersion == oldCluster.ResourceVersion {
+				return
+			}
+			cc.enqueueCluster(newCluster)
+		},
+		DeleteFunc: func(obj interface{}) {
+			// TODO: handle deletion
+		},
+	})
+
+	statefulSetInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: cc.handleObject,
+		UpdateFunc: func(old, new interface{}) {
+			newStatefulSet := new.(*appsv1.StatefulSet)
+			oldStatefulSet := old.(*appsv1.StatefulSet)
+			if newStatefulSet.ResourceVersion == oldStatefulSet.ResourceVersion {
+				return
+			}
+			cc.handleObject(new)
+		},
+		DeleteFunc: cc.handleObject,
+	})
+
+	return cc
+}
+
+// Run starts the ClusterController process loop
+func (cc *ClusterController) Run(threadiness int, stopCh <-chan struct{}) error {
+	defer runtime.HandleCrash()
+	defer cc.queue.ShutDown()
+
+	// 	Start the informer factories to begin populating the informer caches
+	logger.Infof("starting cassandra controller")
+
+	// Wait for the caches to be synced before starting workers
+	logger.Infof("waiting for informers caches to sync...")
+	if ok := cache.WaitForCacheSync(
+		stopCh,
+		cc.clusterListerSynced,
+		cc.statefulSetListerSynced,
+		cc.podListerSynced,
+		cc.serviceListerSynced,
+		cc.serviceAccountListerSynced,
+		cc.roleListerSynced,
+		cc.roleBindingListerSynced,
+	); !ok {
+		return fmt.Errorf("failed to wait for caches to sync")
+	}
+
+	logger.Infof("starting workers")
+	for i := 0; i < threadiness; i++ {
+		go wait.Until(cc.runWorker, time.Second, stopCh)
+	}
+
+	logger.Infof("started workers")
+	<-stopCh
+	logger.Infof("Shutting down cassandra controller workers")
+
+	return nil
+}
+
+func (cc *ClusterController) runWorker() {
+	for cc.processNextWorkItem() {
+	}
+}
+
+func (cc *ClusterController) processNextWorkItem() bool {
+	obj, shutdown := cc.queue.Get()
```

Comment:
this function doesn't look too specific to Cassandra, is this common functionality that could be factored out and help other controllers/operators in the future?  the Cassandra operator is defining some new patterns that the other operators haven't used yet, and we may want to take advantage of them in the future.

edit: this comment applies to a few functions in this file actually, there may be some good generally applicable functionality here

Reply:
This entire file is following the informer - workqueue pattern.
A simple version of this can be found here: https://github.com/kubernetes/sample-controller/blob/master/controller.go
Controllers in the api server also use this pattern.